torch/overrides.py

```diff
@@ -1852,23 +1855,14 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
 
     def __enter__(self):
         old = _get_torch_function_mode()
-        if hasattr(self, "inner"):
-            raise RuntimeError(f"{self} has already been used as a mode. Please use a fresh version or use restore")
-        else:
-            self.inner = old
-            if old is None:
-                self.ancestors = set()
-            else:
-                self.ancestors = self.inner.ancestors.union({self.inner})
+        if self in _cur_torch_function_mode or self is old:
+            raise RuntimeError(f"{self} is already active in the mode stack")
+        _cur_torch_function_mode.append(old)
```

Comment:
Reading through this code it's a bit confusing to distinguish between Python state (the _cur_torch_function_mode) and the C++ state (the _set_torch_function_mode / _get_torch_function_mode). Is it possible to just have a single `_push_mode_stack(mode)`, `_pop_mode_stack()` function that will handle both the python state and c++ state?

Also nit: the names of the C++ state ones (`_set_torch_function_mode` / `_get_torch_function_mode`) should really have `cpp` somewhere in them to avoid confusion